### PR TITLE
Assignment A23

### PR DESCRIPTION
**Pull request by tmmrajesh**

### CODE REVIEW

File: PSICompile/Compiler.cs [hunk 1]
```diff
@@ -17,10 +17,10 @@ public class PSICompiler {
 
          // At this point, the ILCodeGen has generated IL code, save it to 
          // a file and call ILAsm to assemble it:
-         File.WriteAllText ("P:/Output/PSIOutput.il", codegen.S.ToString ());
+         File.WriteAllText ("../Output/PSIOutput.il", codegen.S.ToString ());
```
Comment (tarydon) [inline PSICompile/Compiler.cs]: If you use **../Output**, then the behavior depends on where I execute PSICompile from. So if I am in D:/Test and run the command **P:/Bin/Shell.exe**, then .. will map to D:/. If you start relative to the **Location** of the currently executing assembly, you are safer.

File: PSICompile/ILCodeGen.cs [hunk 2]
```diff
@@ -66,9 +107,36 @@ public class ILCodeGen : Visitor {
...
+
+   public override void Visit (NIfStmt f) {
+      string labl1 = NextLabel (), labl2 = NextLabel ();
+      f.Condition.Accept (this);
+      Out ($"    brfalse {labl1}");
+      f.IfPart.Accept (this);
+      Out ($"    br {labl2}");
+      Out ($"   {labl1}:");
+      f.ElsePart?.Accept (this);
+      Out ($"   {labl2}:");
```
Comment (tarydon) [inline PSICompile/ILCodeGen.cs]: Minor aesthetics: If there is no **else** part, you're generating 2 labels, when one would do.

File: PSICompile/ILCodeGen.cs [hunk 3]
```diff
@@ -121,34 +192,48 @@ public class ILCodeGen : Visitor {
    public override void Visit (NUnary u) {
       u.Expr.Accept (this);
       string op = u.Op.Kind.ToString ().ToLower ();
-      op = op switch { "sub" => "neg", _ => op };
+      op = op switch { "sub" => "neg", "not" => "ldc.i4.0\n    ceq", _ => op };
```
Comment (tarydon) [inline PSICompile/ILCodeGen.cs]: `WriteLn (not 2)` should print -3 (NOT when applied to an integer should do a simple binary not operator), so the opcode generated for not should depend on the type of operand. This is correctly for boolean.

**Review by tarydon — CHANGES_REQUESTED** [review]

File: PSICompile/Compiler.cs (hunk 1 above)
Comment (tmmrajesh) [reply PSICompile/Compiler.cs]: Restored the relative path now. **P:** is the Praxis drive on my computer.

File: PSICompile/ILCodeGen.cs (hunk 3 above)
Comment (tmmrajesh) [reply PSICompile/ILCodeGen.cs]: Fixed it by using the **when**  switch expression modifier.

**Pull request closed without merging**